src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -583,59 +624,76 @@ private void updateModelGraveyardToDelete(
+                        if (exception.isEmpty()) {
+                            step.onResponse(acknowledgedResponse);
+                            return;
+                        }
+                        throw exception.get();
 
-            client.execute(
-                UpdateModelGraveyardAction.INSTANCE,
-                new UpdateModelGraveyardRequest(modelId, isRemoveRequest),
-                ActionListener.wrap(acknowledgedResponse -> {
-                    if (exception.isEmpty()) {
-                        step.onResponse(acknowledgedResponse);
-                        return;
-                    }
-                    throw exception.get();
-
-                }, e -> {
-                    // If it fails to remove the modelId from Model Graveyard, then log the error message
-                    String errorMessage = String.format("Failed to remove \" %s \" from Model Graveyard", modelId);
-                    String failureMessage = String.format("%s%s%s", errorMessage, "\n", e.getMessage());
-                    logger.error(failureMessage);
+                    }, e -> {
+                        // If it fails to remove the modelId from Model Graveyard, then log the error message
+                        String errorMessage = String.format("Failed to remove \" %s \" from Model Graveyard", modelId);
+                        String failureMessage = String.format("%s%s%s", errorMessage, "\n", e.getMessage());
+                        logger.error(failureMessage);
 
-                    if (exception.isEmpty()) {
-                        step.onFailure(e);
-                        return;
-                    }
-                    step.onFailure(exception.get());
-                })
-            );
+                        if (exception.isEmpty()) {
+                            step.onFailure(e);
+                            return;
+                        }
+                        step.onFailure(exception.get());
+                    })
+                );
+            } catch (Exception e) {
+                step.onFailure(e);
+            }
         }
 
         // Clear the metadata of the model for a given modelId
         private void clearModelMetadata(String modelId, StepListener<AcknowledgedResponse> clearModelMetadataStep) {
-            client.execute(
-                UpdateModelMetadataAction.INSTANCE,
-                new UpdateModelMetadataRequest(modelId, true, null),
-                ActionListener.wrap(
-                    clearModelMetadataStep::onResponse,
-                    exception -> removeModelIdFromGraveyardOnFailure(modelId, exception, clearModelMetadataStep)
-                )
-            );
+            // temporary setting thread context to default, this is needed to allow actions on model system index
+            // when security plugin is enabled
+            try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+                client.execute(
```

Comment:
This doesnt touch the model system index. Is this needed?

Reply:
ack